lightly/loss/swav.py

```diff
@@ -0,0 +1,133 @@
+    Q /= sum_Q
+
+    B = Q.shape[1] * world_size
+    K = Q.shape[0] # number of prototypes
+
+    for i in range(iterations):
+        # normalize rows
+        sum_of_rows = torch.sum(Q, dim=1, keepdim=True)
+        if world_size > 1:
+            dist.all_reduce(sum_of_rows)
+        Q /= sum_of_rows
+        Q /= K
+        # normalize columns
+        Q /= torch.sum(Q, dim=0, keepdim=True)
+        Q /= B
+
+    Q *= B
+    return Q.t()
+
+
+class SwaV(nn.Module):
+    """TODO
+    
+    """
+
+    def __init__(self,
+                 temperature: float = 0.1,
+                 sinkhorn_iterations: int = 3,
+                 sinkhorn_epsilon: float = 0.05):
+        super(SwaV, self).__init__()
+        self.temperature = temperature
+        self.sinkhorn_iterations = sinkhorn_iterations
+        self.sinkhorn_epsilon = sinkhorn_epsilon
+
+
+    def subloss(self, z: torch.Tensor, q: torch.Tensor):
+        """Calculates the cross entropy for the SwaV prediction problem.
+
+        Args:
+            z:
+                Similarity of the features and the SwaV prototypes.
+            q:
+                Codes obtained from Sinkhorn iterations.
+
+        Returns:
+            Cross entropy between predictions z and codes q.
+
+        """
+        return - torch.mean(
+            torch.sum(q * F.log_softmax(z / self.temperature, dim=1), dim=1)
+        )
+
+
+    def forward(self,
+                high_resolution_outputs: List[torch.Tensor],
+                low_resolution_outputs: List[torch.Tensor]):
+        """Computes the SwaV loss for a set of high and low resolution outputs.
+
+        Args:
+            high_resolution_outputs:
+                List of similarities of features and SwaV prototypes for the
+                high resolution crops.
+            low_resolution_outputs:
+                List of similarities of features and SwaV prototypes for the
+                low resolution crops.
+
+        Returns:
+            Swapping assignments between views loss (SwaV) as described in [0].
+
+        [0]: SwaV, 2020, https://arxiv.org/abs/2006.09882
+
+        """
+        n_crops = len(high_resolution_outputs) + len(low_resolution_outputs)
+
+        # multi-crop iterations
+        loss = 0.
+        for i in range(len(high_resolution_outputs)):
+
+            # compute codes of i-th high resolution crop
+            with torch.no_grad():
+                q = sinkhorn(
+                    high_resolution_outputs[i].detach(),
+                    iterations=self.sinkhorn_iterations,
+                    epsilon=self.sinkhorn_epsilon
+                )
+
+            # compute subloss for each pair of crops
+            subloss = 0.
+            for v in range(len(high_resolution_outputs)):
+                if v != i:
+                    subloss += self.subloss(high_resolution_outputs[v], q)
+
+            for v in range(len(low_resolution_outputs)):
+                subloss += self.subloss(low_resolution_outputs[v], q)
```

Comment:
Why don't you go over the low_resolution_outputs directly:
```python
for low_resolution_output in low_resolution_outputs:
    subloss += self.subloss(low_resolution_output, q)
```

For the high resolution outputs you can do the same by enumerating over them. Then you can still check the index inequality.

Reply:
I think it's easier to understand the way it is and I wanted to be as close as possible to equation 6 from the paper:
![grafik](https://user-images.githubusercontent.com/65946090/134463822-8e6b134a-30f9-48cc-b0bc-45ff2120b4a5.png)

Reply:
I agree, it's closer to the paper,